cibuildwheel/linux.py

```diff
@@ -129,72 +123,31 @@ def build(project_dir, package_name, output_dir, test_command, test_requires, be
+                try:
+                    process.communicate(stdin_str)
+                except KeyboardInterrupt:
+                    process.kill()
+                    process.wait()
+                return process.returncode
 
-        command = [
-            'docker',
-            'run',
-            '--env',
-            'CIBUILDWHEEL',
-            '--name', container_name,
-            '-i',
-            '-v', '/:/host', # ignored on Circle
-            docker_image,
-            '/bin/bash',
-        ]
-        print('docker command: {}'.format(command))
-        docker_process = subprocess.Popen(
-            command,
-            stdin=subprocess.PIPE,
-            universal_newlines=True,
-        )
-
-        # It will take a bit for docker to get the container up and
-        # running.  Keep trying to copy in the project directory until
-        # it succeeds.  There is a timeout to avoid spinning forever.
-
-        timeout = 180
-        start = monotonic.monotonic()
-        while True:
-            time.sleep(1)
-            command = [
-                'docker',
-                'cp',
-                '{}/.'.format(os.path.abspath(project_dir)),
-                '{}:/project'.format(container_name),
-            ]
-            print('docker command: {}'.format(command))
-            if subprocess.call(command) == 0:
-                break
-
-            if monotonic.monotonic() - start > timeout:
-                raise DockerRunTimeoutError(
-                    'Unable to successfully copy project directory'
-                    ' within {} seconds'.format(timeout)
-                )
-
-        try:
-            docker_process.communicate(bash_script)
-        except KeyboardInterrupt:
-            docker_process.kill()
-            docker_process.wait()
-
-        command = [
-            'docker',
-            'cp',
-            '{}:/output/.'.format(container_name),
-            os.path.abspath(output_dir),
-        ]
-        print('docker command: {}'.format(command))
-        subprocess.check_call(command)
-
-        command = [
-            'docker',
-            'rm',
-            '-v', container_name,
-        ]
-        print('docker command: {}'.format(command))
-        subprocess.check_call(command)
-
-        if docker_process.returncode != 0:
+        container_name = 'cibuildwheel-{}'.format(uuid.uuid4())
+        run_docker(['create',
+                    '--env', 'CIBUILDWHEEL',
+                    '--name', container_name,
+                    '-i',
+                    '-v', '/:/host', # ignored on Circle
+                    docker_image, '/bin/bash'])
+        run_docker(['cp', os.path.abspath(project_dir) + '/.', container_name + ':/project'])
+        script_returncode = run_docker(['start', '-i', '-a', container_name], stdin_str=bash_script)
+        run_docker(['cp', container_name + ':/output/.', os.path.abspath(output_dir)])
+        run_docker(['rm', '-v', container_name])
+
```

Comment:
If we make the error handling to run_docker, this code might be neater as a `try: finally:` block, so that if any of the docker commands fail we still clean up the container before exiting.

e.g.

```
try:
    # docker commands - create, cp, run, cp
finally:
    run_docker(['rm', '-v', container_name])
```

Reply:
OK, will do

Reply:
Actually, if we're already failing anyway, do we really care about cleaning up the docker container?